src/analytics/analytics.service.ts

```diff
@@ -0,0 +1,145 @@
+        const app_version = providedAppVersion || DEFAULT_APP_VERSION;
+        const activationEndEvents = router.events.pipe(filter<ActivationEnd>(e => e instanceof ActivationEnd));
+        const navigationEndEvents = router.events.pipe(filter<NavigationEnd>(e => e instanceof NavigationEnd));
+        this.disposable = navigationEndEvents.pipe(
+            withLatestFrom(activationEndEvents),
+            switchMap(([navigationEnd, activationEnd]) => {
+                const url = navigationEnd.url;
+                const screen_name = activationEnd.snapshot.routeConfig && activationEnd.snapshot.routeConfig.path || url;
+                const params: AngularFireAnalyticsEventParams = {
+                    app_name, app_version, screen_name, url,
+                    firebase_screen_class: undefined,
+                    firebase_screen: screen_name,
+                    outlet: activationEnd.snapshot.outlet
+                };
+                const component = activationEnd.snapshot.component;
+                const routeConfig = activationEnd.snapshot.routeConfig;
+                const loadedConfig = routeConfig && (routeConfig as any)._loadedConfig;
+                const loadChildren = routeConfig && routeConfig.loadChildren;
+                if (component) {
+                    return of({...params, firebase_screen_class: nameOrToString(component) });
+                } else if (loadedConfig && loadedConfig.module && loadedConfig.module._moduleType) {
+                    return of({...params, firebase_screen_class: nameOrToString(loadedConfig.module._moduleType)});
+                } else if (typeof loadChildren === "string") {
+                    // TODO is this an older lazy loading style parse
+                    return of({...params, firebase_screen_class: loadChildren });
+                } else if (loadChildren) {
+                    // TODO look into the other return types here
+                    return from(loadChildren() as Promise<any>).pipe(map(child => ({...params, firebase_screen_class: nameOrToString(child) })));
+                } else {
+                    // TODO figure out what forms of router events I might be missing
+                    return of(params);
+                }
+            }),
+            tap(params => {
+                // TODO perhaps I can be smarter about this, bubble events up to the nearest outlet?
+                if (params.outlet == "primary") {
+                    // TODO do I need to add gtag config for firebase_screen, firebase_screen_class, firebase_screen_id?
+                    //      also shouldn't these be computed in the setCurrentScreen function? prior too?
+                    //      do we want to be logging screen name or class?
+                    analytics.setCurrentScreen(params.screen_name, { global: true })
+                }
+            }),
+            map(params => ({ firebase_screen_id: nextScreenId(params), ...params})),
```

Comment:
@Feiyang1 @hsubox76 what would be the interest level in my moving some the `firebase_*` parameter handling (`firebase_screen_id`, `firebase_previous_id`, `firebase_screen_class`, etc.) into the JS SDK?

I know it breaks the separation of concerns a bit but was non-trivial to get working correctly. It's needed to be able to track user flows in the Firebase Console and BigQuery. The iOS and Android SDKs do this automatically for the developer and their behavior is what I modeled here for Angular. I'll be doing the same treatment for React and Ember routers. I suspect other 3p SDKs will been keen on implementing too.

At the very least we should add information on how to implement in the Firebase docs and add types. FYI @gkaldev @schandel @davideast

Reply:
I don't have a lot of Angular experience so maybe wait for @feiyang1 to take a look but a lot of this looks very Angular-router-specific.  What part/params would be passed to the JS SDK to work with? Do we derive screen_class and previous_id from something or would that be passed in?

Reply:
Hooking into the router will be framework specific of course, my thought would be to add a `logScreenView(this, name: string, additionalParams?: {})` function which would do all the firebase specific event param computation: `firebase_screen_id`, `firebase_previous_screen`, `firebase_screen_class`, `firebase_screen`, `firebase_previous_class`, `firebase_previous_screen`, and `firebase_previous_id`. They're are all undocumented and having to do these on my own to get the most out of Firebase was unexpected. 

For instance `firebase_screen_id` challenging to understand, I couldn't find any resources externally and had to "cheat" and look it up on g3 to find out it's a random INT64 (rather than a hashCode of some kind). I'm not 100% confident I got it right either, hence looping in more DelRel folk.

Reply:
Yeah, if these are standard for other Firebase platforms and we can come up with a sort of general interface I think it would make sense for us to bake some of this handling into the JS SDK.  Let us know if you want to set up a meeting to talk about what we can do with these and pass along your hard-earned research about these params, maybe after you get some more input from other devrels confirming how they work.

Reply:
IIUC, you are asking the JS SDK to keep the state of the previous `screen_view` parameters and attach them to the next `screen_view` request?
Can you share a pointer to how iOS SDK and Android SDK handles it?

Reply:
Yeah, basically I'm thinking all the undocumented `firebase_*` attributes that make the console work should be something we take care of.

I'll send over email now that I'm back from the holiday and we can hop on GVC if additional clarity is needed. That probably would've been a better place for the discussion anyway. 😛

Reply:
My Firebase console is still not showing the information I expect, meaning I did not get this 100% correct yet.